sdks/python/apache_beam/yaml/yaml_transform.py

```diff
@@ -377,6 +378,18 @@ def pipeline_as_composite(spec):
     return dict(spec, name=None, type='composite')
 
 
+def normalize_source_sink(spec):
```

Comment:
This is neat, this should help keep the chain YAML parser simple (everything is just a transform) and still exposing the `source` and `sink` keywords to the user makes it easier for users to structure and write the actual YAMLs. 

We had parsed source and sink transforms differently than others and it got messy trying to enforce best practices across all our different classes of transforms (sources, sinks, custom user transforms, etc) so I think this is better.

Reply:
What if there are multiple sources and/or sinks? Should these take a list? (In chains, for sources we could flatten, for sinks we could write to all.) Or does this get too magical? One can always just put additional ones in the transforms list. (Or let the source itself be a type: Composite with a manual flatten in it.) Any thoughts here?

Reply:
Branching is handled outside the `chain` transform where one names ones inputs explicitly. FWIW, the way we handle different parsing is that we delegate based on the type argument for the non-common parameters. 

I like the idea of deferring magic to later. There is the question of whether we call this `source/sink` vs `sources/sinks` (gets messy to accept both, and we could do singular as multiple is probably the exceptional case) and whether we accept/require a list.

Reply:
Having it plural and requiring a list makes sense to me. I think that swap to multiple sinks and sources will be inevitable. In my experience multiple sources were unusual, but it was actually fairly common to want multiple sinks (typically one for an online store and another to an offline store). 

Requiring a list can be annoying for the YAML user who just wants a singular source and sink, but imo a list of size 1 is actually not too much overhead vs. just a single element.

Reply:
Just to be clear, this YAML resembles what makes sense to me
```
sources:
    - source1
    - source2
transforms:
    - t1
    - t2
    - t3
sinks:
    - s1
```

Reply:
Yeah, the question here is whether source1 would get fed into t1 (with a flatten), or fed into source2 (e.g. something that takes a PCollection of file names and reads the elements) or ???. Similarly with sinks (both of which would behave differently in a composite/pipeline where the inputs are explicitly named vs. a chain). Also, does this make the "normal/easy" case look odd to support the "atypical/branched" one? 

Or would this be better handled by a new, explicit "allof" transform that took a list of transforms and implicitly flattened them (which could be placed in the single source or sink spot)?

Reply:
I see, now I realize I had a misunderstanding of chain. I had thought earlier that the goal was to make chain support everything, but now see there'll be a whole other family of composite transforms that can handle non linear graphs. Side question - can a single pipeline combine a chain with a different composite transform? i.e. define multiple composite transforms that then feeds into a chain all within a single pipeline? From what I've interpreted I think the answer is yes?

I thought about it again and I really like the idea to have one source and sink spot **and** to have that allof transform.
We get a really easy simple path (i.e. linear graph, 1 sink and source), and the allof transform gives the YAML dsl enough expressiveness to expand that to multiple sources & sinks in a clean way. This would already cover basically all of the expressiveness we had at Lyft. I'm also not worried about the implicit flatten in the allof transform because imo forcing the user to use allof and documenting what it does would make that flatten explicit. 

Which way are you leaning?

Reply:
Yes. Chain is restricted to linear chains (where the advantage is that the wiring of inputs to outputs can be inferred by the order, rather than having to be explicitly named and called out). 

I'm leaning towards the single source/sink as well, with an allof (TBD), which is what is implemented here. Also, these are not yet set in stone.

Reply:
Yep! Single source/sink seems like a good start to me